pkg/custompluginmonitor/types/config.go

```diff
@@ -0,0 +1,133 @@
+	// TimeoutString is the global plugin execution timeout string.
+	TimeoutString *string `json:"timeout, omitempty"`
+	// InvokeInterval is the interval at which plugins will be invoked.
+	InvokeInterval *time.Duration `json:"-"`
+	// Timeout is the global plugin execution timeout.
+	Timeout *time.Duration `json:"-"`
+	// MaxOutputLength is the maximum plugin output message length.
+	MaxOutputLength *int `json:"max_output_length, omitempty"`
+	// Concurrency is the number of concurrent running plugins.
+	Concurrency *int `json:"concurrency, omitempty"`
+}
+
+// Custom plugin config is the configuration of custom plugin monitor.
+type CustomPluginConfig struct {
+	// Plugin is the name of plugin which is currently used.
+	// Currently supported: custom.
+	Plugin string `json:"plugin, omitempty"`
+	// PluginConfig is global plugin configuration.
+	PluginGlobalConfig pluginGlobalConfig `json:"pluginConfig, omitempty"`
+	// Source is the source name of the custom plugin monitor
+	Source string `json:"source"`
+	// DefaultConditions are the default states of all the conditions custom plugin monitor should handle.
+	DefaultConditions []types.Condition `json:"conditions"`
+	// Rules are the rules custom plugin monitor will follow to parse and invoke plugins.
+	Rules []*types.CustomRule `json:"rules"`
+}
+
+// ApplyConfiguration applies default configurations.
+func (cpc *CustomPluginConfig) ApplyConfiguration() error {
+	if cpc.PluginGlobalConfig.TimeoutString == nil {
+		cpc.PluginGlobalConfig.TimeoutString = &defaultGlobalTimeoutString
+	}
+
+	timeout, err := time.ParseDuration(*cpc.PluginGlobalConfig.TimeoutString)
+	if err != nil {
+		return fmt.Errorf("Error in parsing global timeout %q. %v", *cpc.PluginGlobalConfig.TimeoutString, err)
+	}
+
+	cpc.PluginGlobalConfig.Timeout = &timeout
+
+	if cpc.PluginGlobalConfig.InvokeIntervalString == nil {
+		cpc.PluginGlobalConfig.InvokeIntervalString = &defaultInvokeIntervalString
+	}
+
+	invoke_interval, err := time.ParseDuration(*cpc.PluginGlobalConfig.InvokeIntervalString)
+	if err != nil {
+		return fmt.Errorf("Error in parsing invoke interval %q. %v", *cpc.PluginGlobalConfig.InvokeIntervalString, err)
+	}
+
+	cpc.PluginGlobalConfig.InvokeInterval = &invoke_interval
+
+	if cpc.PluginGlobalConfig.MaxOutputLength == nil {
+		cpc.PluginGlobalConfig.MaxOutputLength = &defaultMaxOutputLength
+	}
+	if cpc.PluginGlobalConfig.Concurrency == nil {
+		cpc.PluginGlobalConfig.Concurrency = &defaultConcurrency
+	}
+
+	for _, rule := range cpc.Rules {
+		if rule.TimeoutString != nil {
+			timeout, err := time.ParseDuration(*rule.TimeoutString)
+			if err != nil {
+				return fmt.Errorf("Error in parsing rule timeout %+v. %v", rule, err)
+			}
+			rule.Timeout = &timeout
+		}
+	}
+
+	return nil
+}
+
+// Validate verifies whether the settings in CustomPluginConfig are valid.
+func (cpc CustomPluginConfig) Validate() error {
+	if cpc.Plugin != customPluginName {
```

Comment:
Do we really need this? You mentioned that now for different intervals we need to have different plugins, right?

Reply:
Answered this in: https://github.com/kubernetes/node-problem-detector/pull/145/files#r151865145 :)

Reply:
I see. Understood now.